# Homework 1 — Introduction to Machine Learning

ML Zoomcamp 2026 · Dataset: [Car Fuel Efficiency 2026](https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv)

## Setup

In [1]:
import numpy as np
import pandas as pd

## Q1. Pandas version

In [2]:
pd.__version__

'3.0.6'

**Answer:** 3.0.6

## Getting the data

```bash
wget https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv
```

In [3]:
data = pd.read_csv("car_fuel_efficiency_2026.csv")

In [4]:
df = pd.DataFrame(data)

## Q2. Records count

How many records are in the dataset?

In [5]:
len(df)

10000

**Answer:** 10000

## Q3. Fuel types

How many fuel types are presented in the dataset?

In [6]:
df.columns

Index(['model_year', 'origin', 'fuel_type', 'drivetrain', 'num_doors',
       'engine_displacement', 'num_cylinders', 'horsepower', 'vehicle_weight',
       'acceleration', 'fuel_efficiency_mpg'],
      dtype='str')

In [7]:
df.groupby('fuel_type').count()

,model_year,origin,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
fuel_type,,,,,,,,,,
Diesel,2053,2053,2053,2053,2053,2053,1896,2053,2004,2053
Gasoline,6913,6913,6913,6913,6913,6913,6285,6913,6728,6913
Hybrid,1034,1034,1034,1034,1034,1034,942,1034,1004,1034


The groupby gives one row per fuel type: Diesel, Gasoline and Hybrid.

**Answer:** 3

## Q4. Missing values

How many columns in the dataset have missing values?

In [8]:
df.count()

model_year             10000
origin                 10000
fuel_type              10000
drivetrain             10000
num_doors              10000
engine_displacement    10000
num_cylinders          10000
horsepower              9123
vehicle_weight         10000
acceleration            9736
fuel_efficiency_mpg    10000
dtype: int64

The dataset has 10000 rows. Only `horsepower` (9123) and `acceleration` (9736) have fewer non-null values.

**Answer:** 2

## Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

In [9]:
df[df['origin']=='Asia'].fuel_efficiency_mpg.max()

np.float64(41.2)

**Answer:** 41.2

## Q6. Median value of horsepower

1. Find the median of `horsepower`.
2. Find the most frequent value of `horsepower`.
3. Fill the missing values with the most frequent value.
4. Calculate the median again. Has it changed?

In [10]:
df['horsepower'].median()

np.float64(254.0)

In [11]:
df['horsepower'].value_counts().head(n=1)

horsepower
252.0    174
Name: count, dtype: int64

In [12]:
df_na = df['horsepower'].fillna(252)

In [13]:
df_na.median()

np.float64(252.0)

The median went from 254.0 to 252.0 after filling with the most frequent value (252).

**Answer:** Yes, it decreased

## Q7. Sum of weights

Compute $w = (X^T X)^{-1} X^T y$ for the first 7 cars from Asia, using `vehicle_weight` and `model_year`.

In [14]:
X = df[df['origin']=='Asia'][['vehicle_weight','model_year']].head(n=7).values

In [15]:
XTX = X.T.dot(X)
XTX

array([[131950500,  60750580],
       [ 60750580,  28041424]])

In [16]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.96830537e-06, -6.43071025e-06],
       [-6.43071025e-06,  1.39675281e-05]])

In [17]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [18]:
w = XTX_inv.dot(X.T).dot(y)

In [19]:
w.sum()

np.float64(0.36919696904925486)

**Answer:** 0.369